File: aval2/timus1671-anansi_cobweb.py
from sys import stdin, stdout
from collections import deque
n,m = map(int, stdin.next().split(' '))

edges = ' '.join([ stdin.next()[:-1] for _ in xrange(m) ]).split(' ')
_edges = [ (edges[i], edges[i+1]) for i in xrange(0,2*m,2) ]
q = int(stdin.next().strip())
indexes = deque(map(int,stdin.next().split(' ')))
forbidden = set(indexes)

parent = {str(i): str(i) for i in xrange(n+1) }
rank = {str(i): 0 for i in xrange(n+1) }

def find(u):
    if parent[u] != u:
        parent[u] = find(parent[u])
    return parent[u]

def union(u,v):
    ru,rv = find(u), find(v)
    if ru != rv:
        if rank[ru] < rank[rv]:
            parent[ru] = rv
        elif rank[ru] > rank[rv]:
            parent[rv] = ru
        else:
            parent[rv] = ru
            rank[ru] += 1
    return ru != rv

count = n
for i, edge in enumerate(_edges):
	if i+1 not in forbidden:
            if union(*edge):
                count -= 1

answers = deque([str(count)], q)
indexes.popleft()
while indexes:
	if union(*_edges[indexes.pop()-1]):
		count -= 1
	answers.appendleft(str(count))
stdout.write(' '.join(answers))

File: aval2/timus1671-anansi_cobweb.c
#include <stdio.h>
#define MAX 100005

int n,m;
int ex[MAX];
int ey[MAX];
int indexes[MAX];
int allowed[MAX];
int p[MAX];
int sz[MAX];
int ans[MAX];
int q;
int count;

int find(int u){
	int i;
	for(i=u; i != p[i]; i = p[i])
		p[i] = p[p[i]];
	return i;
}

void disj_union(int u, int v){
	int i = find(u);
	int j = find(v);
	int k;

	if(i != j){
		count--;
		if(sz[i] < sz[j])
			p[i] = j;
		else if(sz[j] > sz[i])
			p[j] = i;
		else{
			p[i] = j;
			sz[j]++;
		}
	}
}

int main(){
	int i,j;
	scanf("%d %d", &n, &m);
	count = n;
	for(i=1; i<=m; i++){
		scanf("%d %d", &ex[i], &ey[i]);
		allowed[i] = 1;
	}
	scanf("%d", &q);
	for(i=0; i<q;i++){
		scanf("%d", &indexes[i]);
		allowed[indexes[i]] = 0;
	}
	for(i=1; i<=n; i++){
		p[i] = i;
		sz[i] = 0;
	}
	for(i=1;i<=m;i++){
		if(allowed[i])
			disj_union(ex[i], ey[i]);
	}

	j=1;
	ans[0] = count;
	for(i=q-1; i > 0; i--){
		disj_union(ex[indexes[i]], ey[indexes[i]]);
		ans[j++] = count;
	}

	for(i=q-1; i>0; i--)
		printf("%d ", ans[i]);
	printf("%d\n", ans[i]);
	return 0;
}
